sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -403,55 +410,61 @@ function createFunctionInfo(
-
-        // Note, i can't think of a better way to determine this.  This is particularly hard because
-        // we can't even necessary refer to async function objects here as this code is rewritten by
-        // TS, converting all async functions to non async functions.
-        const isAsyncFunction = func.constructor && func.constructor.name === "AsyncFunction";
+        const isAsyncFunction = await computeIsAsyncFunction(func);
 
         // Ensure that the prototype of this function is properly serialized as well. We only need to do
         // this for functions with a custom prototype (like a derived class constructor, or a function
         // that a user has explicit set the prototype for). Normal functions will pick up
         // Function.prototype by default, so we don't need to do anything for them.
         if (proto !== Function.prototype &&
             !isAsyncFunction &&
-            !isDerivedNoCaptureConstructor(func)) {
+            !await isDerivedNoCaptureConstructorAsync(func)) {
 
-            const protoEntry = getOrCreateEntry(proto, undefined, context, serialize, logInfo);
+            const protoEntry = await getOrCreateEntryAsync(proto, undefined, context, serialize, logInfo);
             functionInfo.proto = protoEntry;
 
-            if (isDerivedClassConstructor) {
-                processDerivedClassConstructor(protoEntry);
+            if (functionString.startsWith("class ")) {
+                // This was a class (which is effectively synonymous with a constructor-function).
+                // We also know that it's a derived class because of the `proto !==
+                // Function.prototype` check above.  (The prototype of a non-derived class points at
+                // Function.prototype).
+                //
+                // they're a bit trickier to serialize than just a straight function. Specifically,
+                // we have to keep track of the inheritance relationship between classes.  That way
+                // if any of the class members references 'super' we'll be able to rewrite it
+                // accordingly (since we emit classes as Functions)
+                await processDerivedClassConstructorAsync(protoEntry);
+
+                // Because this was was class constructor function, rewrite any 'super' references
+                // in it do its derived type if it has one.
                 functionInfo.code = rewriteSuperReferences(funcExprWithName!, /*isStatic*/ false);
             }
         }
 
         // capture any properties placed on the function itself.  Don't bother with
         // "length/name" as those are not things we can actually change.
-        for (const keyOrSymbol of getOwnPropertyNamesAndSymbols(func)) {
-            if (keyOrSymbol === "length" || keyOrSymbol === "name") {
+        for (const descriptor of await getOwnPropertyDescriptors(func)) {
+            if (descriptor.name === "length" || descriptor.name === "name") {
```

Comment:
So this goes from checking names and symbols and just checking names, if I'm understanding this right - is this OK?